utils/in_mem_item_tree/src/primitives/mod.rs

```diff
@@ -0,0 +1,136 @@
+
+/// This represents the node's data item.
+/// Note: the code in this module relies on the fact that the returned ids will never change.
+/// So, implementors of this trait should avoid having mutators for the corresponding fields
+/// (or be extra-careful about accidental mutations).
+pub trait DataItem {
+    type Id: Clone + Copy + Debug + Display + Ord;
+
+    /// Some kind of external object that might be needed by `DataItem`'s implementor to be able
+    /// to produce `Option<&Id>` in `parent_item_id`. E.g. chainstate's `BlockIndex` references its
+    /// parent simply as `Id<GenBlock>`; in order to check whether it's the genesis, one needs access
+    /// to `ChainConfig`, which the classifier may hold in this case.
+    type IdClassifier: Clone + Debug;
+
+    /// Return the id of the item itself.
+    fn item_id(&self) -> &Self::Id;
+
+    /// Return the id of the items's parent or `None` if the item has no parent.
+    /// Note that the exact meaning of `None` may differ depending on the implementor. E.g. if it's
+    /// the chainstate's `GenBlockIndex`, then the tree may contain the genesis itself, so having
+    /// `None` as the parent will mean that this node is the genesis. If it's `BlockIndex`, then
+    /// the tree may not contain the genesis, so having `None` as the parent will mean that this
+    /// node's parent is the genesis.
+    fn parent_item_id(&self, id_classifier: &Self::IdClassifier) -> Option<&Self::Id>;
+}
+
+/// The "flavor" that is used as a generic parameter for the `Arena`.
+pub trait Flavor {
+    /// The modification observer, which is informed about nodes' additions and removals.
+    type ModificationObserver<T: DataItem>: detail::ModificationObserver<T>;
+}
+
+/// The favor that maintains the item-id-to-node-id map in both debug and release builds.
+pub struct WithItemIdToNodeIdMap;
+
+impl Flavor for WithItemIdToNodeIdMap {
+    type ModificationObserver<T: DataItem> = detail::TrackingModificationObserver<T>;
+}
+
+/// The favor that maintains the item-id-to-node-id map only in debug builds.
+pub struct WithDebugOnlyChecks;
+
+impl Flavor for WithDebugOnlyChecks {
+    type ModificationObserver<T: DataItem> = detail::DebugOnlyTrackingModificationObserver<T>;
+}
+
+#[derive(thiserror::Error, Debug, Clone, Eq, PartialEq)]
+pub enum Error {
+    // Note: `indextree::NodeError` doesn't implement Eq/PartialEq, so we store a string instead.
+    #[error("Index tree error: {0}")]
+    IndexTreeError(String),
+    #[error("Index tree utils error: {0}")]
+    IndexTreeUtilsError(#[from] indextree_utils::Error),
+    #[error("Node id {0} is not in arena")]
+    NodeIdNotInArena(NodeId),
+    #[error("Node id {0} is removed")]
+    NodeIsRemoved(NodeId),
+    #[error(
+        "Item {child_item_id} has parent {childs_parent_item_id} instead of expected {expected_parent_item_id}",
+    )]
+    NotParentChild {
+        child_item_id: String,
+        childs_parent_item_id: String,
```

Comment:
children?

Reply:
> children?

It's supposed to be read as `child's`.

Reply:
I would remove the possessive 's', but not a big deal.